### ISSUE TITLE
Questions regarding your training setup

### ISSUE THREAD

**Issue by matthiasfranz197** [JeongHun0716/MMS-LLaMA#3]
Dear Jeong Hun,

again thanks a lot for your work.

I am currently trying to train a MMS-LLaMA model on the 433h LRS3 data from scratch with your training script and it is running without any issues. However, the results are a bit worse than yours. Therefore I first have a few questions about your setup:

- According to the line `valid_subset: test` in the `dataset` part of `src/conf/mms-llama.yaml` you used the test split as validation set during the training. Is this correct or is the current configuration present in the repo not the configuration you used for your model?
- As far as I understood the code in `dataset.py` you add to the train samples babble noise with a probability 0.75 with a randomly drawn SNR level out of {-5, 0, 5, 10, 15, 20}dB and for validation you do not add noise to the samples of the test split. Did I understand that correct?
- Did you change the code for the evaluation in 0dB SNR condition? In the current implementation of `add_noise()` the variable `self.snr_target` is not considered and a random SNR level is selected or did I misunderstood something?

Moreover, I do not have the same hardware available, therefore I used different hardware setups among others:

- 1 A100, 2000 max tokens and update_freq of 4
- 1 H100, 4000 max tokens and update_freq of 2
- 1 A100_80gb, 8000 max tokens and update_freq of 1

The results are all a bit worse, varying from 4.15% to 10.5% with 0dB babble noise.
I observed that in some hypotheses a repetition occurs, e.g.:
```
REF:make it go away
HYP:i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and i can go where i want to go and
``` 
Did you also observe such behavior? Do you have any idea why this happens? I first guessed that it had something to do with the update_freq, but I also observed this behavior with update_freq of 1.

I am looking forward to your answer. Thanks a lot in advance.

Best regards
Matthias

**Comment by JeongHun0716** [comment]
Dear Matthias

Thank you for your interest in this project.
 
- According to the line valid_subset: test in the dataset part of src/conf/mms-llama.yaml you used the test split as validation set during the training. Is this correct or is the current configuration present in the repo not the configuration you used for your model?

Ans) Yes. The configuration file in the repo is identical to the one I used for training.

- As far as I understood the code in dataset.py you add to the train samples babble noise with a probability 0.75 with a randomly drawn SNR level out of {-5, 0, 5, 10, 15, 20}dB and for validation you do not add noise to the samples of the test split. Did I understand that correct?

Ans ) Right. 

- Did you change the code for the evaluation in 0dB SNR condition? In the current implementation of add_noise() the variable self.snr_target is not considered and a random SNR level is selected or did I misunderstood something?

Ans) You’re right. During the initial evaluation, I did override self.snr_levels with snr_target, but that assignment was accidentally removed when I cleaned up the code for release. To restore the intended behavior, set your snr_target as before and then ensure that, whenever you’re evaluating the “test” subset, 

```
if subset == "test":
    snr_levels = [self.snr_target]
```
That way, add_noise() will properly use your specified SNR (e.g., 0 dB) rather than picking a random level.


---
In my opinion, because fairseq uses a bucket sampler to minimize padding by grouping samples of similar duration, the max_tokens parameter may affect training. Have you ever tried a setup with a single A100 GPU, max_tokens=1000, and update_freq=8?

**Comment by matthiasfranz197** [comment]
Dear Jeong Hun,

thanks a lot for your response. 
I have now also run a training with a single A100 GPU, with max_tokens= 1000 and update_freq=8. However, when I run the evaluation with max_tokens=1000 there are still hypotheses with repetitions. When running the evaluation with batch size of 1, I did not observe any repetitions regardless of the model (max_tokens and update_freq used in training) and regardless of the noise conditions. With batch size of 1 the results are also close to your results :)

During my search for the cause of the repetitions, I extracted the `wav_data `of `load_feature()` (in `dataset.py`) before and after the noise addition. I sent the two files for one sample of the test set via mail. Do your audio files after noise addition sound similar?

**Comment by JeongHun0716** [comment]
Thank you for sharing your findings and for sending over the noisy audio samples.

I have a quick follow-up question: when you used the pre-trained model checkpoint and ran inference with max_tokens=1000, did you also observe the repetition issue? 

Additionally, I recall observing a similar phenomenon in another project (zero-avsr). At that time, the issue appeared from differences between versions of the transformers. My experiments were conducted using transformers v4.47.1.

I believe that the more fundamental issue that needs to be addressed is the padding position used during batch training and inference. In my case, I initially used right padding for both training and inference, which caused many repetition issues. When I changed only the inference padding to left padding, the issue almost completely disappeared—and since all the checkpoints had already been trained, I didn’t make any additional code changes. However, if you want a truly complete solution, you might try using left padding for both training and inference.

**Comment by matthiasfranz197** [comment]
Hi Jeong Hun,

Thanks for your quick response. 
With your pre-trained model checkpoint and max_tokens=1000, I did not observe the repetition issue in any inference setup. So this only happens with the models I trained and when padding happens, i.e., when I do not use batch_size=1. I use transformers v4.51.3.

I did not change the part of the padding. So the current implementation does left padding for training and inference in the method `prepare_inputs_labels_for_queries()`in `model.py`, doesn't it?
Did I understand correctly that you initially used right padding for the training of the downloadable checkpoints and then for inference changed the code to left padding?

Best regards
Matthias

**Comment by JeongHun0716** [comment]
I think there might have been some confusion on my part, as I was mixing this up with the zero-avsr project (in that project, we used right padding for training and left padding for inference, and the uploaded code also reflects this). I have just double-checked, and for the checkpoints I uploaded, both training and inference were done using left padding.

So, I believe the cause of your issue could be related to the difference in the transformers library version. Could you try lowering your transformers version to 4.47 and running inference again to see if the problem persists?

**Comment by matthiasfranz197** [comment]
Dear Joeng Hun,

I have a few new insights. As I told you before I did not observe the repetition error with max_tokens=1000, your pre-trained model checkpoint and the transformer v4.51.3. However with max_tokens=3000 and v4.51.3, I had one inference experiment where one hypothesis contains the repetition issue. 

I downgraded to transformers v4.47.1 and so far the repetition issue is gone. For the models that I trained with the v4.51.3 no repetitions occurred so far when I used v4.47.1 for inference. I will switch for the trainings also to v4.47.1. I did not yet repeat all the experiments with v4.47.1, but I think the issue resolved. 

Thank you for your help and your responses
